src/include/com/amazonaws/kinesis/video/webrtcclient/Stats.h

```diff
@@ -281,11 +307,12 @@ typedef struct {
  */
 typedef struct {
     CHAR localId[MAX_STATS_STRING_LENGTH + 1]; //!< Used to look up RTCOutboundRtpStreamStats for the SSRC
-    UINT64 roundTripTime; //!< Estimated round trip time (seconds) for this SSRC based on the RTCP timestamps
-    UINT64 totalRoundTripTime; //!< The cumulative sum of all round trip time measurements in seconds since the beginning of the session
-    UINT64 fractionLost; //!< The fraction packet loss reported for this SSRC
-    UINT64 reportsReceived; //!< Total number of RTCP RR blocks received for this SSRC
-    UINT64 roundTripTimeMeasurements; //!< Total number of RTCP RR blocks received for this SSRC that contain a valid round trip time
+    volatile SIZE_T roundTripTime; //!< Estimated round trip time (milliseconds) for this SSRC based on the RTCP timestamps
```

Comment:
Important! The volatile variables will be used in atomic operations. Those should be machine word aligned. I see the localId and later fractionLost variable that might break the alignment. Perhaps, move the volatile SIZE_Ts up front to ensure proper alignment. The same goes for all of the parameters.

Reply:
is there any pragma or attribute to force 64bit alignment?
as you can see i update stats with atomics, didnt want to add yet another lock for stats, why risk deadlock and/or slowdown when only updating stats. right?

Reply:
I don't think there is a single nice portable way to enforce a 64 bit alignment - you might also have issues on 32 bit platforms. This said, you can always move the volatiles up on top - this is what we do with every structure we have that has members which will be used with atomics.

The other question is whether you can have generic atomics for 64 bit type and the answer is NO. Some of the compilers/platforms/libraries (most don't support) allowing 64bit atomic operations on 32bit internally actually use interlocking. In our case, due to the platform and compiler portability we only assume SIZE_T type atomics operations.

Reply:
> move the volatiles up on top
good idea
> 64bit atomic operations on 32bit internally actually use interlocking
just a thought - do you know how JVM implements it's `AtomicLong` ? might make sense to look into it....

Reply:
I am not entirely sure but it can't be using CPU intrinsics on 32 bit. Likely on nix-es it's wrapped by a futex (similar to synchronized keyword application) and on Windows-es must be using RW locks

Reply:
i cant make myself to use locks

Reply:
c++11 has `std::atomic<uint64>`

Reply:
from AtomicLong.getAndAddLong
```
    public final long getAndAddLong(Object o, long offset, long delta) {
        long v;
        do {
            v = getLongVolatile(o, offset);
        } while (!weakCompareAndSetLong(o, offset, v, v + delta));
        return v;
    }
```
and then
```
UNSAFE_ENTRY(jboolean, Unsafe_CompareAndSwapLong(JNIEnv *env, jobject unsafe, jobject obj, jlong offset, jlong e, jlong x))
  UnsafeWrapper("Unsafe_CompareAndSwapLong");
  Handle p (THREAD, JNIHandles::resolve(obj));
  jlong* addr = (jlong*)(index_oop_from_field_offset_long(p(), offset));
  if (VM_Version::supports_cx8())
    return (jlong)(Atomic::cmpxchg(x, addr, e)) == e;
  else {
    jboolean success = false;
    ObjectLocker ol(p, THREAD);
    if (*addr == e) { *addr = x; success = true; }
    return success;
  }
```

Reply:
@MushMal

Reply:
uses `cmpxchg` if available and mutex otherwise

Reply:
`#include <stdatomic.h>` for plain C

Reply:
There are no 64bit intrinsic atomic operations on 32 bit systems. Some compiler backends allow type promotion to SSE2/SIMD larger register types which on the given system allow atomicity, others, generate spinlock code that handles synchronization. Our atomics can't do that - it's a limitation we need to deal with.

Now, this piece of code is of no concern when it comes to performance so I recommend:

1) For any 64 bit type representing counts, use SIZE_T directly
2) For any duration/time, cast it down to SIZE_T by exposing not 100ns granularity but something more meaningful - say in ms and be explicit about it
3) If there are types that do need 64 bit (which I don't see yet) we need to use a MUTEX

Reply:
ok i'll try to use locks but if it looks too ugly, i'll give atomics another shot

Reply:
Thanks @zhuker . Per my comment, see if you really need the 64 bit type or SIZE_T works OK in which case we could use atomics as is